network/p2p/streams.go

```diff
@@ -42,18 +43,25 @@ type streamManager struct {
 // StreamHandler is called when a new bidirectional stream for a given protocol and peer is opened.
 type StreamHandler func(ctx context.Context, pid peer.ID, s network.Stream, incoming bool)
 
-func makeStreamManager(ctx context.Context, log logging.Logger, h host.Host, handler StreamHandler) *streamManager {
+func makeStreamManager(ctx context.Context, log logging.Logger, h host.Host, handler StreamHandler, allowIncomingGossip bool) *streamManager {
 	return &streamManager{
-		ctx:     ctx,
-		log:     log,
-		host:    h,
-		handler: handler,
-		streams: make(map[peer.ID]network.Stream),
+		ctx:                 ctx,
+		log:                 log,
+		host:                h,
+		handler:             handler,
+		allowIncomingGossip: allowIncomingGossip,
+		streams:             make(map[peer.ID]network.Stream),
 	}
 }
 
 // streamHandler is called by libp2p when a new stream is accepted
 func (n *streamManager) streamHandler(stream network.Stream) {
+	if stream.Conn().Stat().Direction == network.DirInbound && !n.allowIncomingGossip {
```

Comment:
As I understand it, streams and connections can have different directions. Why is the connection direction important here instead of the stream direction (`stream.Stat().Direction`)?

I'm concerned this still allows gossip to be pulled from a node by any peers it has an outbound connection with.

Reply:
This is a fair question. With this EnableGossipService flag we inherit concept of "client-server" so that incoming and outgoing connections (made in `DialPeersUntilTargetCount` until `GossipFanout`) can be distinguished. This is the only fact we can control because streams can be opened by either side and only peerID order defines should it be kept or not.
This basically allows us to implement a "server" with EnableGossipService=true to accept gossip streams and "client" nodes with "EnableGossipService=false" to continue receiving and sending gossip traffic because this client client node made connection, not someone else connected to them. This aims to protect the p2p listening port from be a gossip port, and it serves this purpose.

Reply:
I thought about it more, libp2p prevents duplicate connections if there is already A->B connection, it will not make B->A.